### ISSUE TITLE
Address repohascommitafter search filter performance bottleneck

### ISSUE THREAD

**Issue by kzh** [sourcegraph/sourcegraph#6745]
This issue is to track fixing the `repohascommitafter` search filter performance bottleneck. At the moment, it is preemptively generating a list of repository that satisfy the commit after date. Then it searches over that list. Instead, we can do this on the fly to avoid extraneous network/gitserver requests.

/cc @sourcegraph/search-platform

**Comment by beyang** [comment]
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.11 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

**Comment by tsenart** [comment]
@kzh: Can you expand the issue description with what would need to be done? If this is no longer relevant, please close.

**Comment by uwedeportivo** [comment]
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.13 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

**Comment by kzh** [comment]
The original method I proposed in the issue description didn't plan out. I'm currently working on another way to speed up this filter. The speed bottleneck stems from having to make two gitserver calls (one for resolving revision and one for checking if the repository has a commit past a certain date) per repository when this filter is active. To avoid having to do this, we can store the last commit date in a column in the `repo` table. This field will be refreshed when the repository is updated in `repo-updater`. We'll just specify in the SQL statement to filter past a certain date. This should be a substantial speedup.

This is how the archive and fork filters work right now as well.
(https://sourcegraph.com/github.com/sourcegraph/sourcegraph@55b1f110585ab10a337811b6dbb7a90a57f278d7/-/blob/cmd/frontend/graphqlbackend/search.go#L502-512)
```go
repos, err = db.Repos.List(ctx, db.ReposListOptions{
	OnlyRepoIDs:     true,
	IncludePatterns: includePatterns,
	ExcludePattern:  unionRegExps(excludePatterns),
	// List N+1 repos so we can see if there are repos omitted due to our repo limit.
	LimitOffset:  &db.LimitOffset{Limit: maxRepoListSize + 1},
	NoForks:      op.noForks,
	OnlyForks:    op.onlyForks,
	NoArchived:   op.noArchived,
	OnlyArchived: op.onlyArchived,
})
```
Thoughts on this approach? I recall wanting to avoid storing git data in postgres since it would be redundant. cc @sourcegraph/core-services

**Comment by kzh** [comment]
The original method I proposed in the issue description didn't plan out. I'm currently working on another way to speed up this filter. The speed bottleneck stems from having to make two gitserver calls (one for resolving revision and one for checking if the repository has a commit past a certain date) per repository when this filter is active. To avoid having to do this, we can store the last commit date in a column in the `repo` table. This field will be refreshed when the repository is updated in `repo-updater`. We'll just specify in the SQL statement to filter past a certain date. This should be a substantial speedup.

This is how the archive and fork filters work right now as well.
(https://sourcegraph.com/github.com/sourcegraph/sourcegraph@55b1f110585ab10a337811b6dbb7a90a57f278d7/-/blob/cmd/frontend/graphqlbackend/search.go#L502-512)
```go
repos, err = db.Repos.List(ctx, db.ReposListOptions{
	OnlyRepoIDs:     true,
	IncludePatterns: includePatterns,
	ExcludePattern:  unionRegExps(excludePatterns),
	// List N+1 repos so we can see if there are repos omitted due to our repo limit.
	LimitOffset:  &db.LimitOffset{Limit: maxRepoListSize + 1},
	NoForks:      op.noForks,
	OnlyForks:    op.onlyForks,
	NoArchived:   op.noArchived,
	OnlyArchived: op.onlyArchived,
})
```
Thoughts on this approach? I recall wanting to avoid storing git data in postgres since it would be redundant. cc @sourcegraph/core-services

**Comment by keegancsmith** [comment]
Lets stick to codehost metadata in the repos table. If we want to store computed values, lets stick them in their own table / redis. We can then use those computed values to speed up filtering.

Introducing something which computes something every time repo-updater thinks a repo has changed is a very useful bit of infrastructure. I would be pro doing this if we do it in a way which targets this use case but done in a way where there are incremental steps to generalise it. Note: repo-updater doesn't always know about a repo changing (EnsureHasRevision on git-server being the main one, or some customers switching off repo-updater updating repos). But it has an upper-bound on discovering that.

Alternatively we can more incrementally just introduce a cache on the "latest" commit date. That cache can be used to know if a repo is past a certain date. For dates before we would have to double check since a new commit may have arrived.

**Comment by tsenart** [comment]
> Lets stick to codehost metadata in the repos table.

Can you share some rationale for this? I'm fine with storing this in a separate table and then joining, but would like to understand your thinking behind it.

**Comment by keegancsmith** [comment]
right now everything written to the repo row comes directly from a `repos.Source` (ie directly from a codehost API). I think it would be harder to understand data if we introduce columns which are updated on a different cadence.

I imagine everything we introduce that is like this (say rev for HEAD, or latest commit date) will be updated async.

I also don't know if the correct change is doing it as a SQL join, rather than in our application logic doing the filtering.

And finally I am always hesitant introducing higher write rates in postgres, especially on something which is as large and has as many indexes as our repo table.

**Comment by kzh** [comment]
> Introducing something which computes something every time repo-updater thinks a repo has changed is a very useful bit of infrastructure. I would be pro doing this if we do it in a way which targets this use case but done in a way where there are incremental steps to generalise it.

I like this idea a lot. AFAIK, the existing pieces of architecture that are similar to this are zoekt indexserver (pull) and the LSIF indexer (push), but these are fairly specialized cases. A more general approach would also bring us closer to becoming a "code metastore" (I recall this briefly being discussed in the past). We would be able to scrape and retain a lot more information from a codebase. Aside from this issue, #2587 (language inventory) would greatly benefit from this as well. This seems RFC worthy.

> Note: repo-updater doesn't always know about a repo changing

Ah right. I didn't consider this. Is it the case that the gitserver repository eventually updates from the code host without user intervention? If so, it might be okay to use stale data from up to ten minutes ago.

> Alternatively we can more incrementally just introduce a cache on the "latest" commit date. That cache can be used to know if a repo is past a certain date. For dates before we would have to double check since a new commit may have arrived.

Do you mean an in-memory map of some sorts (or redis) that is updated on the repo-updater cadence? I'm worried about the initial search speed. I'm assuming this filter is usually used in a global search query.

**Comment by keegancsmith** [comment]
> > Note: repo-updater doesn't always know about a repo changing
> 
> Ah right. I didn't consider this. Is it the case that the gitserver repository eventually updates from the code host in the background without being user triggered? If so, it might be okay to use stale data from up to ten minutes ago.

The main use-case is a user browsing on there code-host with the sourcegraph extension. They are browsing code for a commit we don't have. We have logic in our gitserver API which allows setting a commit that should exist. When the commit does not exist we do a "git fetch" before running the git command. This bit of functionality we have tried to remove in the past, since it would be nice to only have remote commands triggered by repo-updater.

> > Alternatively we can more incrementally just introduce a cache on the "latest" commit date. That cache can be used to know if a repo is past a certain date. For dates before we would have to double check since a new commit may have arrived.
> 
> Do you mean using an in-memory map of some sorts (or redis)? I'm worried about the initial search speed slowness from the gitserver network calls. I'm assuming this filter is usually used in a global search query.

Redis. The incremental thing to do to avoid building a whole piece of infra around this is to just prime the cache somewhere in repo-updater where we ping/check repos?

Maybe taking a step back there is something else we can do? IE store the commit/author date of HEAD in zoekt? IE in the case of using indexed search we stick to using indexed search. Otherwise we fallback to our slower/JIT search infra which is slower (and potentially has a cache). Just trying to think of ways to avoid making a huge project out of this :)

**Comment by keegancsmith** [comment]
We chatted over VC about this. The decision we made that relying on zoekt isn't really good enough. Adding in a cache will be too unreliable. So working on the original idea of some infrastructure which computes this information on repository change makes sense to do.

However, we should probably see if there are any quicker wins to unblock customers experiencing issues with repohascommitafter at a larger scale.

**Comment by tsenart** [comment]
@keegancsmith: This could be a good customer issue for @rvantonder this iteration? Or are there other open issues that are more important coming from the distribution team's list?

**Comment by keegancsmith** [comment]
https://github.com/sourcegraph/sourcegraph/issues/8568 is a duplicate and indicates we also give bad results when the filtering fails / timesout. We should probably address the UI issues first.



> @keegancsmith: This could be a good customer issue for @rvantonder this iteration? Or are there other open issues that are more important coming from the distribution team's list?

Yes this would be interesting to tackle. We have general timeout issues in search which are more important, but I haven't properly triaged all of them yet.

**Comment by rvantonder** [comment]
Caught up on the discussion. Copying over context from @keegancsmith's comments in #8568 

> the feature as currently implemented has poor performance characteristics (we check each repository to see if it passes the filter). We need to move to an indexed approach. Doing this properly is likely out of scope for 3.14. However, I am sure we can adjust the meaning slightly to bring it back into scope. If instead of finding any commit after a certain date, would just finding all repositories with HEAD past a certain date suffice? That way it would quite neatly fit into our current indexed search architecture.

> Either way I'll see how we can make the reporting better on the search page / see if there is some lightweight caching we can do.

Seems there isn't a clear, robust solution path here without a new system. Agree developing a long term solution seems too large a scope to tackle this iteration.

> just finding all repositories with HEAD past a certain date suffice?

Sounds like this could be a good interim direction. One question re #8568 is reproducing the behavior to have an indication of of whether this approach helps the underlying issue (and how much), or if there is a chance something else is going on. Will grab a call with @keegancsmith.

**Comment by kzh** [comment]
Some last thoughts: Eric F's work on batch request support in gitserver reminded me of this. This is another potential quick solution. Most of the slowness comes from all the gitserver requests (two per repository, one for resolving head and one for the actual filtering git command). The requests themselves process and return quickly. However, they stack up.
- https://sourcegraph.com/github.com/sourcegraph/sourcegraph@e33dd28df3b487a16d98dd85a593f08d3e778f3e/-/blob/cmd/frontend/graphqlbackend/search.go#L644
- https://sourcegraph.com/github.com/sourcegraph/sourcegraph@e33dd28df3b487a16d98dd85a593f08d3e778f3e/-/blob/cmd/frontend/graphqlbackend/search.go#L563:1

All of this can be avoided by adding a gitserver endpoint for "batch repohascommitafter" that accepts a list of repositories and returns a filtered list based on the repohascommiterafter argument. This avoids the unnecessary overhead coming from network requests. An indexed solution would still be better.

**Comment by tsenart** [comment]
@rvantonder: Please triage.

**Comment by uwedeportivo** [comment]
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.14 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

**Comment by keegancsmith** [comment]
@rvantonder do you think you will have time to work on this in 3.16? Have you done more work on this? Could this be something good for any one else in search team to pick in 3.16?

**Comment by rvantonder** [comment]
It's not clear to me how to go ahead with this. Summarizing the discussion below, my take is that _if_ this is addressed, it needs to be done as a whole, and that's a significant task.  I don't think this should top other things I have on the stack. The last discussion fizzled out on short-win proposals:

> Keegan: I think the feature is poorly thought out to be honest. I'd change it to instead be based on HEAD's commit timestamp. Then the features becomes much more tractable to do in a performant way. I would like to do that since I suspect that satisfies most users needs and we can tractably implement that sooner rather than later.

> Stephen: I like Keegan's idea, how is that different from what we do today? It seems strange to me that basing this off HEAD would be substantially different? Is it trying to check for all commits, effectively?
I think the users who care about this just want to "search or do not search unmaintained repositories" basically, so HEAD only would be fine I imagine

> Stephen: Perhaps we can take this opportunity to make the syntax nicer as well? maintained:true/false (or stale:true/false)and then describe this in the docs as a hueristic which currently looks at HEAD's timestamp but may use improved heuristics in the future

> Rijnard: I think we already just check against HEAD in most cases. It depends on what resolveRepositories does, but if we just send over a bunch of repos without a revision it'll resolve to HEAD and then hit the HasCommitAfter call:
https://sourcegraph.com/github.com/sourcegraph/sourcegraph@e33dd28df3b487a16d98dd85a593f08d3e778f3e/-/blob/cmd/frontend/graphqlbackend/search.go#L540-563

> Rijnard: I very much dislike the syntax. I agree the feature is poorly thought out. This to me means it should get the proper attention with an RFC and proposals (I am overcommitted and can't do this myself). I think changing the syntax should be considered as part of the whole picture (what does true/false mean, etc). 

---

In a separate thread, I discussed rough ideas for syntax, but the _real_ issue with this functionality and issue is the perf part, and what users actually want to do with things like 'stale' or 'unmaintained', and whether that can inform other ways to improve perf. So I'll elide the syntax discussion for now.

**Comment by rvantonder** [comment]
Putting in backlog, feel free to bring it out if there are better ways to move forward.

**Comment by tsenart** [comment]
@keegancsmith: I believe we need to capture the general git metadata indexing idea in a separate issue or RFC. It may fit well in search OKRs for Q2.

**Comment by keegancsmith** [comment]
> @keegancsmith: I believe we need to capture the general git metadata indexing idea in a separate issue or RFC. It may fit well in search OKRs for Q2. 

Agreed. An RFC is called for here. Addressing in Q2 also makes sense,
since it is a bit of infra that will be an enabler for other
features. Will tackle next week.

**Comment by camdencheek** [comment]
This may be closeable since we now list repos in a streaming manner and in parallel. I'll leave it to the platform team to make the call on closing though

**Comment by camdencheek** [comment]
Clsoing as done since we now search search repos iteratively
